vector/v.fill.holes/main.c

```diff
@@ -0,0 +1,224 @@
+    G_add_keyword(_("fill"));
+    G_add_keyword(_("exterior"));
+    G_add_keyword(_("ring"));
+    G_add_keyword(_("perimeter"));
+    options.module->description =
+        _("Fill holes in areas by keeping only outer boundaries");
+
+    options.input = G_define_standard_option(G_OPT_V_INPUT);
+
+    options.field = G_define_standard_option(G_OPT_V_FIELD);
+
+    options.cats = G_define_standard_option(G_OPT_V_CATS);
+
+    options.where = G_define_standard_option(G_OPT_DB_WHERE);
+
+    options.output = G_define_standard_option(G_OPT_V_OUTPUT);
+
+    if (G_parser(argc, argv))
+        exit(EXIT_FAILURE);
+
+    Vect_check_input_output_name(options.input->answer,
+                                 options.output->answer, G_FATAL_EXIT);
+    Vect_set_open_level(2);
+
+    if (1 >
+        Vect_open_old2(&input, options.input->answer, "",
+                       options.field->answer))
+        G_fatal_error(_("Unable to open vector map <%s>"),
+                      options.input->answer);
+
+    /* Check if old vector is 3D. We should preserve 3D data. */
+    if (Vect_is_3d(&input))
+        open3d = WITH_Z;
+    else
+        open3d = WITHOUT_Z;
+
+    /* Set error handler for input vector map */
+    Vect_set_error_handler_io(&input, NULL);
+
+    /* Open new vector for reading/writing */
+    if (0 > Vect_open_new(&output, options.output->answer, open3d)) {
+        G_fatal_error(_("Unable to create vector map <%s>"),
+                      options.output->answer);
+    }
+
+    /* Set error handler for output vector map */
+    Vect_set_error_handler_io(NULL, &output);
+
+    int field = Vect_get_field_number(&input, options.field->answer);
+
+    if (field <= 0 && options.cats->answer)
+        G_fatal_error(_("Option %s cannot be combined with %s=%s"),
+                      options.cats->key, options.field->key,
+                      options.field->answer);
+    if (field <= 0 && options.where->answer)
+        G_fatal_error(_("Option %s cannot be combined with %s=%s"),
+                      options.where->key, options.field->key,
+                      options.field->answer);
+
+    /* Copy header and history data from old to new map */
+    Vect_copy_head_data(&input, &output);
+    Vect_hist_copy(&input, &output);
+    Vect_hist_command(&output);
+
+    // Set category constraint.
+    struct cat_list *constraint_cat_list = NULL;
+
+    if (field > 0)
+        constraint_cat_list =
+            Vect_cats_set_constraint(&input, field, options.where->answer,
+                                     options.cats->answer);
+
+    /* Create and initialize struct's where to store points/lines and categories */
+    struct line_pnts *points = Vect_new_line_struct();
+    struct line_cats *area_cats = Vect_new_cats_struct();
+    struct line_cats *empty_cats = Vect_new_cats_struct();
+
+    struct ilist *all_cats = Vect_new_list();
+    struct ilist *field_cats = Vect_new_list();
+
+    plus_t num_areas = Vect_get_num_areas(&input);
+
```

Comment:
create an index for lines already copied to the output, needed to write out boundaries only once, initialized to zero
```suggestion
plus_t num_lines = Vect_get_num_lines(&input);
int *line_written_out = G_calloc(num_lines + 1, sizeof(int));
```

Reply:
Thanks. This was my main misunderstanding. The topology for areas is build automatically from a pile of centroids and boundaries, but I need to handle those directly instead of just dealing with areas.